UTM Conversion Utility

dthomas1953 opened the pull request:
### What does this PR do?
It adds capability to convert UTM coordinates to latitudes and longitudes. The relevant script is ph5:utilities:utmtolatlong.py, and has been integrated with entry_points.py, with the INGESTION group.
There are two ways to use it: interactive:
```
# Interactive method:
#    enter -e Easting, -n Northing, -z Zone (1-60), -s Hemisphere (S, N)
# example:
# utmtolatlong -e 540947 -n 1361594 -z 58 -s S
# which results in output:
# easting=540947.0, northing=1361594.0, zone=58, hemisphere=S
# Lon = 166.7381664, Lat = -77.8156740
```
and Batch Mode. For the file utmdata.txt:
```
easting, northing, zone, hemisphere
540947,1361594,58,S
322916,3771967,13,N
```
running the command `utmtolatlong -i utmdata.txt -o mylatlong.txt` 
creates the output file mylatlong.txt:
```
Longitude, Latitude
166.7381664, -77.8156740
-106.9190960, 34.0734958
```

### Relevant Issues?
"read UTM or UTM to lat/lon converter" #301

### Checklist
- [X] All tests pass. (see ph5.utilities.tests.test_utmtolatlong.py)
- [X] Any new or changed features have are documented.
- [x] Changes have been added to `CHANGELOG.txt` . *oops, will have to do that.*

nick-iris commented:
@dthomas1953 Doesn't the `ph5.core.cs2cs` module already have methods for doing this?

dthomas1953 commented:
@nick-iris , I took a look at 'ph5.core.cs2cs.py`.  There are a couple of show-stopping problems with trying to use this module to address #301.

* There is no capability for entering easting, northing, zone & hemisphere for conversion to lat/long coordinates. You can input X,Y,Z coordinates, but #301 is related to Fairfield Nodes, which record coordinates as UTM Eastings and Northings. 

* The code is broken.  
```>>> a = cs2cs.utm2geod(58, 'WGS84', 540497, 1361594, 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cs2cs.py", line 63, in utm2geod
    return lat, lon, Z
NameError: global name 'lat' is not defined
```
The error is in this line: `lon, lan = p(X, Y, inverse=True)` 
which should read `lon, lat = p(X, Y, inverse=True)`

But even with that fixed, there is no way to enter easting/northing/etc.

The Fairfield Nodes do not use (X,Y,Z) coordinates.

dthomas1953 commented:
Found some flake8 problems with entry_points, will clobber those and do a new pull request.

nick-iris commented:
@dthomas1953 Did you mean to close this pull request? We still need to review

nick-iris commented:
You can just push your flake8 changes to this PR

dthomas1953 commented:
Okay, stand by.

dthomas1953 commented:
flacated entry_points, updated CHANGELOG.TXT, next stop, review?

nick-iris commented on ph5/utilities/utmtolatlong.py:
```diff
@@ -0,0 +1,167 @@
+            return
+        hemisphere = str(args.side).upper()
+        if hemisphere != "S" and hemisphere != "N":
+            print ("Error, zone must be either N or S for northern/southern \
```
You should be using a the logger, not printing to stdout.

nick-iris replied:
You should be using a the logger, not printing to stdout.

nick-iris commented on ph5/utilities/utmtolatlong.py:
```diff
@@ -0,0 +1,167 @@
+
+def main():
+
+    desc = 'Convert UTM locations to longitudes, latitudes, \
```
I'm noticing these `\^M` carriage-return characters. Are you working on a Windows machine?

Please set the file encoding to UTF-8 and line-endings for new files to Unix in your IDE. This is so that text files are saved in a format that is not specific to the Windows OS and most easily shared across heterogeneous developer desktops.

nick-iris replied:
I'm noticing these `\^M` carriage-return characters. Are you working on a Windows machine?

Please set the file encoding to UTF-8 and line-endings for new files to Unix in your IDE. This is so that text files are saved in a format that is not specific to the Windows OS and most easily shared across heterogeneous developer desktops.

nick-iris commented on ph5/utilities/utmtolatlong.py:
```diff
@@ -0,0 +1,167 @@
+PROG_VERSION = '2019.212'
+
+
+class utm_conversions:
```
Class names should always be upper case. I suggest changing this class name to `UTMConversions`

nick-iris commented on ph5/utilities/utmtolatlong.py:
```diff
@@ -0,0 +1,167 @@
+    parser.add_argument("-s", "--side", help="Side=Hemisphere, N or S")
+
+    args = parser.parse_args()
+    # print ("args:", args)
```
Remove commented out code

nick-iris commented on ph5/utilities/utmtolatlong.py:
```diff
@@ -0,0 +1,167 @@
+    try:
+        fli = len(args.infile)
+        flo = len(args.outfile)
+    except (RuntimeError, TypeError, NameError, ValueError):
```
Why is the try/except necessary?

dthomas1953 replied:
I will get the class names nd comments cleaned up, and will set encoding to UTF-8. I'm using Gedit for Mac, I didn't see the ^M characters.
The try/except is necessary if the user uses the inline `-e Easting -n Northing etc.` method, which is distinct from the batch file method. In the inline case, the variables for infile and outfile are never defined, and testing them for length raises an error. They are only defined if user uses the `-i ifile -o outfile` option. Thus the try/catch.
Going to a funeral, will do another push later this morning.

nick-iris replied:
> In the inline case, the variables for infile and outfile are never defined, and testing them for length raises an error. They are only defined if user uses the -i ifile -o outfile option.

Instead of catching these errors you should explicitly check if `args.infile` and `args.outfile` exist. If they exist then take their length, else set `fli` and `flo` to zero.

nick-iris replied:
> I'm using Gedit for Mac, I didn't see the ^M characters.

You can see these carriage returns in GitHub (https://github.com/PIC-IRIS/PH5/pull/321/files).

dthomas1953 commented:
@nick-iris , I added `import logging` at the top of utmtolatlong.py, followed by `LOGGER = logging.getLogger(__name__)`.
In the script, I replaced the print() command (prints info to screen) with items like this:
```python
        if easting < 0.0 or northing < 0.0:
            warn = "Error, easting and northing must both be"
            warn += " positive, but you used {0} and {1}"
            LOGGER.info(warn.format(easting, northing))
            return
```

The program runs (doesn't crash), but no text is sent to the screen. How do I get a screen display like the following for command `utmtolatlong -e 540947 -n 1361594 -z 58 -s S`, using LOGGER?
```
easting=540947.0, northing=1361594.0, zone=58, hemisphere=S
Lon = 166.7381664, Lat = -77.8156740
```

nick-iris commented:
Just use the LOGGER for reporting errors or warnings. For example `LOGGER.error(warn.format(easting, northing))` instead of LOGGER.info.

You can print the main output to stdout. I don't think you need the line with the `***` characters though.

dthomas1953 commented:
Ran out of time today, we had to prep some field laptops and etc. Will bang out an update for Friday.

nick-iris commented:
@dthomas1953 There's no rush. Better to have everything be well written.

dthomas1953 commented on ph5/utilities/utmtolatlong.py:
```diff
@@ -0,0 +1,167 @@
+# This script can be used to convert UTM coordinates to Latitude/Longitude
```
Good suggestions, I'll ponder these next week.

dsentinel commented:
@ascire-pic, what file formats are you using or do you envision using?
What is the process for updating coordinates for a node?

> PH5 being able to read in UTM in addition to lat/lon for location information or having an integrated UTM converter would help make metadata fixes for node locations easier.

What do you mean by integrated?

dsentinel commented:
https://github.com/PIC-IRIS/PH5/blob/894357e0cbcfb2e4fcf2b9e0485ec7d4b25d9009/ph5/utilities/segd2ph5.py#L981

Here is the code that currently does the conversion.
Lets make this generic as Dave has done, and have all code use the one version.
Perhaps in `core.ph5utils.py`

Then an entry point CLI can wrap it and work on the file format / use case @ascire-pic needs.
Place the cli in `utilities`

ascire-pic commented:
@dsentinel I needed to update the coordinates for a node PH5 where the array tables were automatically generated by segdtoph5. However, the PI gave me the updated locations in UTM instead of lat/lon because that is what the node server uses. 

To update the coordinates, I used keftoph5 and keftocsv to get a csv-format version of the array tables and then replaced the lat/lon columns with the corrected locations. To convert from UTM given to me to lat/lon I found a random converter online and used that, but it would have been easier if there was an easy-to-use UTM converter included in the PH5 software suite. I envisioned either feeding in a text file with columns for easting, northing and UTM zone, or having UTM as an input option in noven.

dthomas1953 commented:
@dsentinel , @nick-iris : Did the refactoring per Lloyd, ran flake8 on new files, tests pass, I think this is ready to go. I could use advice on "failing check" above.

derick-hess commented:
@dthomas1953 the test is failing during ph5/core/tests/test_ph5api.py

It is failing the first time it tries to read lat and lon from a Node in the generated ph5 archive. It expects the ['location/X/value_d'] to be  -105.405489539  and ['location/Y/value_d'] to be 47.6790599342 when it reads the array table for this node, but they are now both being written as 0.0

So I think something is now wrong with segd2ph5 when it tries to convert utm to lat, lon and write it in the array tables.

dthomas1953 commented:
Good clue, @derick-hess , I'll check it out.

dthomas1953 commented:
Hey, is there a way to capture test results to a file? When I try
`python runtest.py > output.txt`, it only saves a few lines. And, no help trying `python runtest.py | more`

I was able to ctrl-break out of the tests and peruse the problem. It might be that, for hemisphere (north/south), the original method had it go to north if undefined, and I had changed it to None for the new way. Trying that fix now.

derick-hess commented:
The script that creates the PH5 test data set runs

`segdtoph5 -n master.ph5 -U 13N -r ../segd/3ch.fcnt`

So it should be setting the hemisphere correctly. It was before, and segdtoph5 was properly detecting north and south.

You can just run an individual test script directly if you want.

dthomas1953 commented:
I'm running just the TestPH5API test, that reproduces the error. Unfortunately, the hemisphere of north vs None didn't help. Tracking it down!
```
(ph5) [dthomas@New-OS-Image-9 PH5]$ cat mytest.py
import unittest
from ph5.core.tests.test_ph5api import TestPH5API
suite = unittest.TestLoader().loadTestsFromTestCase(TestPH5API)
runner = unittest.TextTestRunner(verbosity=3)
print runner.run(suite)

```
output:
```
test_array_t (ph5.core.tests.test_ph5api.TestPH5API) ... FAIL
test_channels (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_close_ph5 (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_cut (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_das_t (ph5.core.tests.test_ph5api.TestPH5API) ... FAIL
test_event_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_experiment_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_get_availability (ph5.core.tests.test_ph5api.TestPH5API) ... [2019-08-14 18:25:28,494] - ph5.core.ph5api - WARNING: No Das table found for 12345
ok
test_get_extent (ph5.core.tests.test_ph5api.TestPH5API) ... [2019-08-14 18:25:28,662] - ph5.core.ph5api - WARNING: No Das table found for xxxx
ok
test_index_t (ph5.core.tests.test_ph5api.TestPH5API) ... FAIL
test_load_ph5 (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_mix_ins (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_offset_t (ph5.core.tests.test_ph5api.TestPH5API) ... [2019-08-14 18:25:29,132] - ph5.core.ph5api - WARNING: Couldn't get offset.
ok
test_read_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_receiver_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_response_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_sort_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_textural_cut (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_time_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_trace_padding (ph5.core.tests.test_ph5api.TestPH5API) ... ok

======================================================================
FAIL: test_array_t (ph5.core.tests.test_ph5api.TestPH5API)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dthomas/PH5/ph5/core/tests/test_ph5api.py", line 130, in test_array_t
    self.assertEqual(channel[0]['location/X/value_d'], -105.405489539)
AssertionError: 0.0 != -105.405489539

======================================================================
FAIL: test_das_t (ph5.core.tests.test_ph5api.TestPH5API)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dthomas/PH5/ph5/core/tests/test_ph5api.py", line 330, in test_das_t
    self.assertEqual(1, len(table))
AssertionError: 1 != 5

======================================================================
FAIL: test_index_t (ph5.core.tests.test_ph5api.TestPH5API)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dthomas/PH5/ph5/core/tests/test_ph5api.py", line 249, in test_index_t
    self.assertEqual(11, len(self.ph5API_object.Index_t['rows']))
AssertionError: 11 != 35

----------------------------------------------------------------------
Ran 20 tests in 2.265s

FAILED (failures=3)
<unittest.runner.TextTestResult run=20 errors=0 failures=3>
```

derick-hess commented:
Just in case, make sure you deleted the old ph5 test data set and recreated it with create_ph5.sh before rerunning the test so it generates a new ph5 archive with the code changes

dthomas1953 commented:
@derick-hess Thanks, I hadn't done that!

dsentinel commented:
> Just in case, make sure you deleted the old ph5 test data set and recreated it with create_ph5.sh before rerunning the test so it generates a new ph5 archive with the code changes

Doesn't / shouldn't the test create it. Many of the tests print that it will build it if not found. However, I don't think they all work.

derick-hess commented:
> > Just in case, make sure you deleted the old ph5 test data set and recreated it with create_ph5.sh before rerunning the test so it generates a new ph5 archive with the code changes
> 
> Doesn't / shouldn't the test create it. Many of the tests print that it will build it if not found. However, I don't think they all work.

It can be done that way. The tests that are testing actually building PH5 do build a small data set and then delete is afterwards. obspytoPH5 and metadatatoph5 do this, as should  tests for <anything>toph5 when those tests are written.

You could have test_ph5api,  	test_ph5availability.py, test_ph5toms.py, test_ph5tostationxml.py and all the others build the ph5 in their setup and tear it back down. That would add a significant amount of time to running tests though. Also thought about just putting a test ph5 archive in a data directory, but didn't want to add a bunch of data files to the repository, but that is also an option.

dthomas1953 commented:
Good news, the hemisphere default worked, once I started testing from scratch.

```
Your branch is up-to-date with 'origin/utm_util'.

(ph5) [dthomas@New-OS-Image-9 PH5]$ python ../mytest.py
test_array_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_channels (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_close_ph5 (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_cut (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_das_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_event_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_experiment_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_get_availability (ph5.core.tests.test_ph5api.TestPH5API) ... [2019-08-14 20:09:30,798] - ph5.core.ph5api - WARNING: No Das table found for 12345
ok
test_get_extent (ph5.core.tests.test_ph5api.TestPH5API) ... [2019-08-14 20:09:30,967] - ph5.core.ph5api - WARNING: No Das table found for xxxx
ok
test_index_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_load_ph5 (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_mix_ins (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_offset_t (ph5.core.tests.test_ph5api.TestPH5API) ... [2019-08-14 20:09:31,443] - ph5.core.ph5api - WARNING: Couldn't get offset.
ok
test_read_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_receiver_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_response_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_sort_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_textural_cut (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_time_t (ph5.core.tests.test_ph5api.TestPH5API) ... ok
test_trace_padding (ph5.core.tests.test_ph5api.TestPH5API) ... ok

----------------------------------------------------------------------
Ran 20 tests in 2.284s

OK
```

derick-hess commented:
I'm not sure that actually fixed it. Segdtoph5 is passing the UTM zone as 13N so it shouldn't be getting to that final else statement. If you change the create_ph5 script to 13S do the coordinates change to correctly be in the southern hempisphere?

dthomas1953 commented:
I'll give it a try.

dthomas1953 commented:
Other tasks intervene, I'll get back to this tomorrow. Thanks for the tips.

derick-hess commented:
Weird, the tests definitely failed and it said pass. That needs to be looked at
You deleted all the test data out of of the repository. That test data is needed to run the tests

dthomas1953 commented:
I realized I had deleted too much test data yesterday, will give it another go today.

dsentinel commented:
@dthomas1953, This branch is now the same as master. You can re-implement your changes.

dthomas1953 commented:
@dsentinel , Thanks!

dthomas1953 commented:
I redid it, found the flaw in segd2ph5.py, utmcsptolatlon() that was messing up tests.

Old and new tests pass, new stuff is flake8ed,  can we merge this now?

dthomas1953 commented:
OK, I see the problem:
```
=====================================================================
FAIL: test_array_t (ph5.core.tests.test_ph5api.TestPH5API)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/PIC-IRIS/PH5/ph5/core/tests/test_ph5api.py", line 130, in test_array_t
    self.assertEqual(channel[0]['location/X/value_d'], -105.405489539)
AssertionError: 0.0 != -105.405489539
```
I'll see what happened there.

dthomas1953 commented:
I found why the test was failing, somehow I reverted one of the fixes I made earlier.

Should be good to merge, now.

dsentinel commented:
I force pushed. After the mess of backtracking.
Everything is clean now.

dthomas1953 commented:
Looks good.

dthomas1953 commented:
I downloaded a clean version of utm_util. runtests.py is much simplified. The UTM utilities run OK, and all the tests passed. BUT, when I ran runtests.py,I didn't see the test cases for UTM. 

The UTM test file is there, and I can run it interactively:
```python
>>> import unittest
>>> from ph5.core.ph5utils import UTMConversions
>>> from ph5.utilities.tests.test_utmtolatlong import TestUTMconversion
>>> suite = unittest.TestLoader().loadTestsFromTestCase(TestUTMconversion)
>>> runner = unittest.TextTestRunner(verbosity=3)
>>> runner.run(suite)
test_for_correct_type (ph5.utilities.tests.test_utmtolatlong.TestUTMconversion) ... ok
test_for_correct_value (ph5.utilities.tests.test_utmtolatlong.TestUTMconversion) ... ok
test_is_valid_utm_conversion (ph5.utilities.tests.test_utmtolatlong.TestUTMconversion) ... ok
test_is_valid_utm_conversion_south (ph5.utilities.tests.test_utmtolatlong.TestUTMconversion) ... ok
----------------------------------------------------------------------
Ran 4 tests in 0.231s

OK
<unittest.runner.TextTestResult run=4 errors=0 failures=0>
```
Why aren't these results appearing when I run runtests.py?

When I look at that interactively, looks like the info is there alright:
```python
>>> import unittest
>>> ts = unittest.defaultTestLoader.discover('.')
>>> print ts
. . .
<unittest.suite.TestSuite tests=[<unittest.suite.TestSuite tests=[<ph5.utilities.tests.test_utmtolatlong.TestUTMconversion testMethod=test_for_correct_type>, <ph5.utilities.tests.test_utmtolatlong.TestUTMconversion testMethod=test_for_correct_value>, <ph5.utilities.tests.test_utmtolatlong.TestUTMconversion testMethod=test_is_valid_utm_conversion>, <ph5.utilities.tests.test_utmtolatlong.TestUTMconversion testMethod=test_is_valid_utm_conversion_south>]>]>
```

So the tests should be in the test suite, why don't they appear? (I looked.)

dthomas1953 commented:
Lloyd and I had a look, it was just a verbosity issue. With verbosity set=3 in runtests.py, the results I was expecting indeed showed up.

dsentinel commented:
There seems to be an issue with `from pyproj import Transformer` not existing in pyproj 1.9, but is ok in 2.2

dthomas1953 commented:
@nick-iris , @dsentinel I think this is ready to merge. Can you review?
Thanks, Dave

dthomas1953 commented:
@nick-iris , @dsentinel this is finally ready to be merged. Has some UTM utilities (conversions in both directions) that the data group has been waiting for.

dthomas1953 commented:
When I create and run tests on my repository, the results are vastly different than shown here in Travis. Why?

[ph5-create-test.txt](https://github.com/PIC-IRIS/PH5/files/4119604/ph5-create-test.txt)

dsentinel commented:
@dthomas1953 , You had a conflict merging from master. I resolved it for ya.
Now it seems as though there is a proper error in the api.

dthomas1953 commented:
@dsentinel thanks, I'll have a look.

dthomas1953 commented:
@dsentinel , @nick-iris I found the fly in the ointment, this should be OK to merge finally.

dthomas1953 commented:
@dsentinel , UTM class removed, loops simplified, comments cleaned up, functions renamed, I think it's ready this time.

dsentinel commented on ph5/core/ph5utils.py:
```diff
@@ -68,6 +71,76 @@ def get_n_i(self, sensor_keys, datalogger_keys):
+
+    az, baz, dist = g.inv(lon0, lat0, lon1, lat1)
+
+    if dist:
```
If this conditional is to test ==0, then `if dist == 0`

dsentinel commented on setup.py:
```diff
@@ -54,7 +54,7 @@
                       'numpy',
                       'numexpr',
-                      'pyproj',
+                      'pyproj>=2',
```
?

dsentinel replied:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

dthomas1953 commented:
@dsentinel , I will tend to these.  I took the ```pyproj>=2``` out, dunno how that crept back in.

I plan to go ahead and add a function to do the UPS (Univ. Polar Stereographic) conversions also, since I have the formula from Fairfield, and since we're expecting south pole data soon.

dthomas1953 commented:
@dsentinel @nick-falco I've added UPS (Univ. Polar Stereographic) support, cleaned up function names, replaced print commands with logger commands, and the other requested changes. Hoping this is a go for merging this time.

dthomas1953 commented on ph5/core/ph5utils.py:
```diff
@@ -68,6 +71,105 @@ def get_n_i(self, sensor_keys, datalogger_keys):
+    return (northing, easting, utm_zone, hemisphere)
+
+
+def lat_lon_elev_to_utm(lat, lon, elev):
```
@dsentinel , Earlier you said ```consistent with ordering of arguments and return tuples i.e. lat, long and northing, easting```
I endeavored to order the arguments consistently, and now have uniform calling of Y first (latitude, northing), X second (longitude easting), and that's why the calls now include ```always_xy = False```. These functions are now working correctly, both for new functions (utm, ups) and for existing functions that needed utm conversions.
I don't understand what you mean by 
```I'm not sure how the tests are passing: Using Transformer.from_crs(...always_xy=False).transform() takes arguments and returns in order lat, long and northing, easting```, can you explain what the issue is here?

I will remove the un-necessary function, rename the TSPC function, etc.

I am not pinning pyproj version at this time, that was a temporary occurrence that is **not** in the current version. It's resolved already. :/

dthomas1953 commented on ph5/core/ph5utils.py:
```diff
@@ -68,6 +71,105 @@ def get_n_i(self, sensor_keys, datalogger_keys):
+                                       always_xy=False)
+    easting, northing = transformer.transform(lat, lon)
+    return (northing, easting, utm_zone, hemisphere)
+
```
OK

dthomas1953 commented on ph5/core/ph5utils.py:
```diff
@@ -68,6 +71,105 @@ def get_n_i(self, sensor_keys, datalogger_keys):
+        epsgroot = "327"
+    elif hemisphere == "N":
+        epsgroot = "326"
+    epsg_utm = "EPSG:" + epsgroot + str(zone)
```
Looks good, will implement this syntax.

dsentinel commented on ph5/core/ph5utils.py:
```diff
@@ -68,6 +71,105 @@ def get_n_i(self, sensor_keys, datalogger_keys):
+    transformer = Transformer.from_crs(epsg_wgs84,
+                                       epsg_utm,
+                                       always_xy=False)
+    easting, northing = transformer.transform(lat, lon)
```
ORDER!!!!!!!!!!!!

dsentinel commented:
@dthomas1953 
Please halt work on this
